File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int r = -1;
    public int diameterOfBinaryTree(TreeNode root) {
        helper(root);
        return r;
    }
    public int helper(TreeNode root){
        if(root == null) return -1; //base case
        int left = 1 + helper(root.left); //longest path to the left
        int right = 1 + helper(root.right); //longest path to right
        r = Math.max(r , (left+right)); // update the longes diameter found 
        return Math.max(left , right); // return the longest path that passes through this node
    }
}

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        def dfs(root):
            if not root:
                return 0, 0
            left_curr, left_max = dfs(root.left)
            right_curr, right_max = dfs(root.right)
            path = left_curr + right_curr 
            return max(left_curr, right_curr) + 1, max(left_max, right_max, path)
        return dfs(root)[1]